utils/stall_free.hh

```diff
@@ -56,14 +58,93 @@ void merge_to_gently(std::list<T>& list1, const std::list<T>& list2, Compare com
     }
 }
 
-template<class T>
-seastar::future<> clear_gently(std::list<T>& list) {
-    return repeat([&list] () mutable {
-        if (list.empty()) {
-            return seastar::stop_iteration::yes;
-        }
-        list.pop_back();
-        return seastar::stop_iteration::no;
+template <typename T>
+concept GentlyClearable = requires (T x) {
+    { x.clear_gently() } -> std::same_as<future<>>;
+};
+
+template <typename T>
+concept SmartPointer = requires (T x) {
+    { x.reset() } -> std::same_as<void>;
+    { x.get() } -> std::same_as<typename T::element_type*>;
+};
+
+template <typename T>
+concept VectorLike = requires (T x, size_t n) {
+    { x.empty() } -> std::same_as<bool>;
+    { x.back() } -> std::same_as<typename T::value_type&>;
+    { x.pop_back() } -> std::same_as<void>;
+};
+
+template <typename T>
+concept Container = !VectorLike<T> && requires (T x) {
+    { x.empty() } -> std::same_as<bool>;
+    { x.erase(x.begin()) } -> std::same_as<typename T::iterator>;
+};
+
+template <typename T, std::size_t N>
+future<> clear_gently(std::array<T, N>&a) noexcept;
+
+template <Container T>
+future<> clear_gently(T& c) noexcept;
+
+template <GentlyClearable T>
+future<> clear_gently(T& o) noexcept(noexcept(o.clear_gently())) {
+    return o.clear_gently();
+}
+
+template <typename T>
+future<> clear_gently(T&) noexcept {
+    return make_ready_future<>();
+}
+
+template <SmartPointer Ptr>
+future<> clear_gently(Ptr& o) {
+    return clear_gently(*o).finally([&o] {
+        return o.reset();
```

Comment:
I'm curious, why is this o.reset() important? Why can't we just leave the pointer pointing to a now-empty vector? In fact, isn't it normal that someone clear()s a vector and then continue to write to it new elements?

Reply:
It's a good question.  Clearing a vector destroys all entries in it but leaves its allocation behind so that new objects can be emplaced in it.  A smart pointer is akin to a vector of size 1 hence I destroyed its contents.  It is also intended to be used internally when a value to be cleared gently is wrapped in a smart pointer inside a container.  With that in mind, leaving the pointer intact will work too since the container will typically destroy it when erasing the element (or calling `pop_back` for stack/vector containers)

Reply:
Fixed in v6
no reset of smart pointers, just call to internal::clear_gently(*p)
also, clearing_gently of `SharedPointer` checks the use_count to be 1 for gently clearing the pointed object.